src/builder_state.rs

```diff
@@ -679,57 +663,32 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
     ) -> Self {
         let txns_in_queue: HashSet<_> = tx_queue.iter().map(|tx| tx.commit).collect();
         BuilderState {
-            included_txns: HashSet::new(),
-            included_txns_old: HashSet::new(),
-            included_txns_expiring: HashSet::new(),
+            num_nodes,
             txns_in_queue,
             built_from_proposed_block,
-            decide_receiver,
-            da_proposal_receiver,
-            qc_receiver,
             req_receiver,
-            da_proposal_payload_commit_to_da_proposal: HashMap::new(),
-            quorum_proposal_payload_commit_to_quorum_proposal: HashMap::new(),
-            tx_receiver,
             tx_queue,
             global_state,
-            builder_commitments: HashSet::new(),
-            total_nodes: num_nodes,
             maximize_txn_capture_timeout,
             base_fee,
             instance_state,
-            txn_garbage_collect_duration,
-            next_txn_garbage_collect_time: Instant::now() + txn_garbage_collect_duration,
             validated_state,
+            included_txns: RotatingSet::new(txn_garbage_collect_duration),
+            da_proposal_payload_commit_to_da_proposal: HashMap::new(),
+            quorum_proposal_payload_commit_to_quorum_proposal: HashMap::new(),
+            builder_commitments: HashSet::new(),
+            decide_receiver: receivers.decide.activate_cloned(),
+            da_proposal_receiver: receivers.da_proposal.activate_cloned(),
+            qc_receiver: receivers.quorum_proposal.activate_cloned(),
+            tx_receiver: receivers.transactions.activate_cloned(),
         }
     }
     pub fn clone_with_receiver(&self, req_receiver: BroadcastReceiver<MessageType<TYPES>>) -> Self {
-        // Handle the garbage collection of txns
-        let (
-            included_txns,
-            included_txns_old,
-            included_txns_expiring,
-            next_txn_garbage_collect_time,
-        ) = if Instant::now() >= self.next_txn_garbage_collect_time {
-            (
-                HashSet::new(),
-                self.included_txns.clone(),
-                self.included_txns_old.clone(),
-                Instant::now() + self.txn_garbage_collect_duration,
-            )
-        } else {
-            (
-                self.included_txns.clone(),
-                self.included_txns_old.clone(),
-                self.included_txns_expiring.clone(),
-                self.next_txn_garbage_collect_time,
-            )
-        };
+        let mut included_txns = self.included_txns.clone();
+        included_txns.rotate();
```

Comment:
👍 This is nice.

Reply:
Yeah, this was the most distracting bit when I started reading the codebase for the first time, so I wanted it to abstract it away.